proto/spacemesh/types.proto

```diff
@@ -0,0 +1,162 @@
+}
+
+enum TransactionProcessingStatus {
+    UNDEFINED = 0;
+    REJECTED = 1;               // rejected pre STF processing. e.g. bad data
+    INSUFFICIENT_FUNDS = 2;     // rejected pre STF processing by funds check.
+    CONFLICTING = 3;            // rejected pre STF due to conflicting counter.
+    PENDING = 4;                // pending processing by STF
+    //PROCESSING = 5;
+    //APPROVED = 6;
+    PROCESSED = 7;              // processed. Results are in the tx receipt.
+}
+
+enum TransactionType {
+    COIN = 0;            // simple coin transaction
+    ATX = 1;             // smesher activation transaction
+    APP = 2;             // smart contract method
+    DEPLOY = 3;          // deploy app from templlate
+    DEPLOY_TEMPLATE = 4; // deploy app template code to mesh
+}
+
+enum TransactionResult { // results of processing a transaction by the STF
+    RUNTIME_EXCEPTION = 0;              // app code exception
+    INSUFFICIENT_GAS_PROVIDED = 1;      // out of gas
+    INSUFFICIENT_BALANCE = 2;           // failed due to sender insufficient funds
+    EXECUTED = 3;                      // executed w/o error
+}
+
+message TransactionReceipt {
+    TransactionId id = 1;  // the source transaction
+    TransactionResult result = 2;   // tx processing result
+    uint64 gas_used = 3;            // gas units used by the transaction (gas price in tx)
+    Amount fee = 4;                 // transaction fee charged for the transaction
+}
+
+message AppEvent {                     // an event emitted from an app instance
+    TransactionId transaction_id = 1;  // the transaction that called the code
+    string message = 2;                // the event's string emitted from code
+}
+
+message TransactionId {
+    bytes id = 1;
+}
+
+message SmesherId {
+    bytes id = 1;
+}
+
+message TransactionFee { // = gas * gas_price
+    uint64 gas = 1;
+    uint64 gas_price = 2;
+}
+
+// Simple coin transaction
+message Transaction {
+    TransactionType type = 1;
+    TransactionId id = 2; // hash
+    AccountId sender = 3;
+    TransactionFee fee = 4;
+    uint64 timestamp = 5;  // shouldn't this be part of the event envelope?
+    uint64 layer_id = 6;
+    AccountId receiver = 7; // depending on tx type
+    Amount amount = 8; // amount of coin transfered in this tx by sender
+    TransactionProcessingStatus status = 9;
+    uint64 counter = 10;    // tx counter aka nonce
+
+    bytes data = 11; // binary payload - used for app, deploy, atx and spwan transactions
+    SmesherId smesher_id = 12;    // used in atx only
+    TransactionId prev_atx = 13;  // previous ATX. used in atx.
+}
+
+message Reward {
+    uint64 layer = 1;
+    Amount total = 2;
+    Amount layer_reward = 3;
+    uint64 layer_computed = 4; // layer number of the layer when reward was computed
+    // tx_fee = total- layer_reward
+    AccountId coinbase = 5; // account awarded this reward
+    SmesherId smesher = 6;  // it will be nice to always have this in reward events
```

Comment:
Is there such thing as a `SmesherId`? didn't we say this is going away or something?